rust nm: Preserve existing configurations

cathay4t opened the pull request:
Only update user desired configure and preserving previous existing
configurations.

Enabled the integration test `test_reapply_preserve_ip_config.py` for
rust code.

almusil commented on rust/src/lib/nm/connection.rs:
```diff
@@ -251,39 +213,55 @@ pub(crate) fn get_port_nm_conns<'a>(
     } else {
-        iface_name.to_string()
+        let mut new_nm_conn_set = NmSettingConnection::new();
+        let conn_name = if iface.iface_type() == InterfaceType::OvsBridge {
```
Please use `match` instead of ifs.

almusil commented on rust/src/lib/nm/bridge.rs:
```diff
@@ -1,17 +1,24 @@
+        cur_nm_br_set.clone()
+    } else {
+        NmSettingBridge::new()
+    };
```
Can be:
```
let mut nm_br_set = nm_conn.bridge.as_ref().cloned().unwrap_or_default();
```

almusil commented on rust/src/lib/nm/connection.rs:
```diff
@@ -58,78 +59,39 @@ pub(crate) fn iface_to_nm_connections(
-    };
-    let nm_ctrl_type = if let Some(ctrl_type) = &base_iface.controller_type {
-        Some(iface_type_to_nm(ctrl_type)?)
+    let mut nm_conn = if let Some(exist_nm_conn) = exist_nm_conn {
```
Can be:
```
let mut nm_conn = exist_nm_conn.cloned().unwrap_or_default();
```

almusil commented on rust/src/lib/nm/connection.rs:
```diff
@@ -58,78 +59,39 @@ pub(crate) fn iface_to_nm_connections(
+    gen_nm_ip_setting(iface, &mut nm_conn)?;
+    gen_nm_wired_setting(iface, &mut nm_conn);
 
     if let Interface::OvsBridge(ovs_br_iface) = iface {
```
Please use `match` instead.

almusil commented on rust/src/lib/nm/connection.rs:
```diff
@@ -251,39 +213,55 @@ pub(crate) fn get_port_nm_conns<'a>(
+    iface: &Interface,
+    nm_conn: &mut NmConnection,
+) -> Result<(), NmstateError> {
+    let mut nm_conn_set = if let Some(cur_nm_conn_set) = &nm_conn.connection {
```
Can be:
```
let mut nm_conn_set = nm_conn.connection.as_ref().cloned().unwrap_or_default();
```

almusil replied:
Missed comment?

cathay4t replied:
This one cannot use `unwrap_or_default` as we need to setup many stuff for new connection instead of default.

almusil commented on rust/src/lib/nm/connection.rs:
```diff
@@ -251,39 +213,55 @@ pub(crate) fn get_port_nm_conns<'a>(
+        None
+    };
 
+    let nm_ctrl_type =
```
Can be:
```
let nm_ctrl_type = iface
        .base_iface()
        .controller_type
        .as_ref()
        .map(iface_type_to_nm)
        .transpose()?;
```

ffmancera replied:
Why is the transpose required here? Probably I am missing something..

ffmancera replied:
Why is the trnspose() required? If I am not wrong, this should return only one element.. Am I missing something? Thanks!

almusil replied:
The `map()` here makes `Option<Result<T>>`, so doing transpose 'flips' it into `Result<Option<T>>` on which you can use `?`.

ffmancera replied:
Oh that is true. Thanks for explaining!

almusil commented on rust/src/lib/nm/ip.rs:
```diff
@@ -20,15 +21,22 @@ pub(crate) fn iface_ipv4_to_nm(
     };
-    let mut nm_setting = NmSettingIp::new();
+
+    let mut nm_setting = if let Some(s) = &nm_conn.ipv4 {
```
Can be:
```
    let mut nm_setting = nm_conn.ipv4.as_ref().cloned().unwrap_or_default();
```

almusil commented on rust/src/lib/nm/ip.rs:
```diff
@@ -57,8 +65,42 @@ pub(crate) fn iface_ipv6_to_nm(
         NmSettingIpMethod::Disabled
     };
-    let mut nm_setting = NmSettingIp::new();
+    let mut nm_setting = if let Some(s) = &nm_conn.ipv6 {
```
Can be:
```
let mut nm_setting = nm_conn.ipv6.as_ref().cloned().unwrap_or_default();
```

almusil commented on rust/src/lib/nm/ovs.rs:
```diff
@@ -176,17 +177,28 @@ fn remove_ovs_port_name_prefix(nm_port_name: &str) -> &str {
-    let mut nm_ovs_port_set = NmSettingOvsPort::new();
+    exist_nm_conn: Option<&NmConnection>,
+) -> Result<NmConnection, NmstateError> {
+    let mut nm_conn = if let Some(exist_nm_conn) = exist_nm_conn {
```
Can be:
```
let mut nm_conn = exist_nm_conn.cloned().unwrap_or_default();
```

almusil commented on rust/src/lib/nm/ovs.rs:
```diff
@@ -176,17 +177,28 @@ fn remove_ovs_port_name_prefix(nm_port_name: &str) -> &str {
+    iface.base = base_iface;
+    gen_nm_conn_setting(&Interface::Unknown(iface), &mut nm_conn)?;
+
+    let mut nm_ovs_port_set =
```
Can be:
```
let mut nm_ovs_port_set = nm_conn.ovs_port.as_ref().cloned().unwrap_or_default();
```

almusil commented on rust/src/lib/nm/ovs.rs:
```diff
@@ -222,10 +234,15 @@ pub(crate) fn get_ovs_port_name(
-    let mut nm_ovs_br_set = NmSettingOvsBridge::new();
+    nm_conn: &mut NmConnection,
+) {
+    let mut nm_ovs_br_set = if let Some(s) = &nm_conn.ovs_bridge {
```
Can be:
```
 let mut nm_ovs_br_set = nm_conn.ovs_bridge.as_ref().cloned().unwrap_or_default();
```

almusil replied:
Missed comment?

cathay4t replied:
Fixed.

almusil commented on rust/src/lib/nm/ovs.rs:
```diff
@@ -238,13 +255,15 @@ pub(crate) fn create_nm_ovs_br_set(
-) -> NmSettingOvsIface {
-    let mut nm_ovs_iface_set = NmSettingOvsIface::new();
+pub(crate) fn gen_nm_ovs_iface_setting(nm_conn: &mut NmConnection) {
+    let mut nm_ovs_iface_set = if let Some(s) = &nm_conn.ovs_iface {
```
Can be:
```
let mut nm_ovs_iface_set = nm_conn.ovs_iface.as_ref().cloned().unwrap_or_default();
```

almusil commented on rust/src/lib/nm/wired.rs:
```diff
@@ -0,0 +1,25 @@
+    iface: &Interface,
+    nm_conn: &mut NmConnection,
+) {
+    let mut nm_wired_set = if let Some(cur_nm_wired_set) = &nm_conn.wired {
```
Can be: 
```
let mut nm_wired_set = nm_conn.wired.as_ref().cloned().unwrap_or_default();
```

almusil reviewed (commented):
Maybe the pattern would deserve some macro but that can be always improved later on.

ffmancera commented on automation/run-tests.sh:
```diff
@@ -179,6 +179,7 @@ function run_tests {
           pytest \
             $PYTEST_OPTIONS \
             tests/integration/static_ip_address_test.py \
+            tests/integration/test_reapply_preserve_ip_config \
```
This is incorrect I think you meant -k option here. Please, check https://github.com/nmstate/nmstate/runs/4222094965?check_suite_focus=true#step:6:1600

cathay4t replied:
I should use `preserve_ip_config_test.py` instead of test name. Fixed.

cathay4t commented:
Patch has issue on UUID controller reference for OVS. Working on it.

cathay4t commented:
@almusil @ffmancera Patch ready for another round of review. Thanks!

almusil reviewed (commented):
Other than that it looks good

ffmancera reviewed (approved):
LGTM!

cathay4t committed:
rust nm: Preserve existing configurations

Only update user desired configure and preserving previous existing
configurations.

Enabled the integration test `test_reapply_preserve_ip_config.py` for
rust code.

Signed-off-by: Gris Ge <fge@redhat.com>
